packages/loopring_v3/contracts/impl/libexchange/ExchangeDeposits.sol

```diff
@@ -66,31 +66,34 @@ library ExchangeDeposits
         timestamp = request.timestamp;
     }
 
-    function depositTo(
+    function deposit(
         ExchangeData.State storage S,
-        address recipient,
+        address from,
+        address to,
         address tokenAddress,
         uint96  amount,  // can be zero
         uint    additionalFeeETH
         )
         external
     {
-        require(recipient != address(0), "ZERO_ADDRESS");
+        require(from != address(0), "ZERO_ADDRESS");
+        require(to != address(0), "ZERO_ADDRESS");
+        require(S.isAuthorized(from), "UNAUTHORIZED");
         require(S.areUserRequestsEnabled(), "USER_REQUEST_SUSPENDED");
         require(getNumAvailableDepositSlots(S) > 0, "TOO_MANY_REQUESTS_OPEN");
 
         uint16 tokenID = S.getTokenID(tokenAddress);
         require(!S.tokens[tokenID].depositDisabled, "TOKEN_DEPOSIT_DISABLED");
 
-        uint24 accountID = S.getAccountID(recipient);
+        uint24 accountID = S.getAccountID(to);
         ExchangeData.Account storage account = S.accounts[accountID];
 
         // Total fee to be paid by the user
         uint feeETH = additionalFeeETH.add(S.depositFeeETH);
 
         // Transfer the tokens to this contract
         transferDeposit(
-            account.owner,
+            from,
```

Comment:
so we need to change this method such that if msg.sender == from, we use safeTransfer, otherwise, we use safeTransferFrom. Right?

Reply:
No, the funds are always need to be transferred from the `from` address, never from the contract itself (which safeTransfer would need). This change also fjust for the meta tx, `from` can authorize the deposit directly himself (`from == msg.sender`) or indirectly using the Relayer using an ECDSA signture/gas station network (`relayer == msg.sender`) implemented in a module.